src/selectors/expandThoughts.ts

```diff
@@ -85,13 +103,30 @@ function expandThoughts(state: State, path: Path | null, { depth = 0, returnCont
   const childrenUnfiltered = showContexts
     ? getContexts(state, headValue(simplePath))
     : getAllChildren(state, context) as (Child | ThoughtContext)[]
-  const children = state.showHiddenThoughts
+
+  const expansionPathContext = pathToContext(expansionPath)
+
+  // Note: A path that is ancestor of the expansion path or expansion path itself should always be visible.
+  const visibleChildren = state.showHiddenThoughts
     ? childrenUnfiltered
-    : childrenUnfiltered.filter(child => !isFunction(childValue(child, showContexts)))
+    : childrenUnfiltered.filter(child => {
+      const value = strip(childValue(child, showContexts))
+      const childPath = unroot([...path || [], { ...child, value }])
+      const childContext = pathToContext(childPath)
+
+      /** Check of the path is the ancestor of the expansion path. */
+      const isAncestor = () => isDescendant(childContext, expansionPathContext)
+
+      /** Check if the path is equal to the expansion path. */
+      const isExpansionPath = () => equalArrays(childContext, expansionPathContext)
+
+      const shouldHide = isFunction(value) && !isAncestor() && !isExpansionPath()
+      return !shouldHide
+    })
```

Comment:
This is a lot of work within a filter predicate. For every child in the expansion tree it is calling multiple functions of `O(path.length)`: `pathToContext`, `isDescendant`, `equalArrays`.

Is there a way to reuse work from previous recursive calls and improve the running time complexity of this? For example, if a thought A was not a descendant of the expansion base path in its iteration, then we know that the child of thought A is not a descendant of the expansion base path without having to call `isDescendant`.

Reply:
Yes. I will try to come up with a solution.